swift-sdk/Internal/EmbeddedSessionManager.swift

```diff
@@ -13,67 +13,81 @@ public class EmbeddedSessionManager {
+            session.impressions[index].displayDuration = Double(displayDurationStr) ?? 0
+        }
+        let _ = IterableAPI.track(embeddedSession: session)
+        session.isActive = false
     }
     
-    public func pauseImpression(impressionId: String) {
-        if var trackingImpression = currentlyTrackingImpressions[impressionId], trackingImpression.tracking {
+    public func pauseImpression(messageId: String) {
+        if var trackingImpression = currentlyTrackingImpressions[messageId], trackingImpression.tracking {
             let currentTime = Date()
             let elapsedTime = currentTime.timeIntervalSince(trackingImpression.startTime)
             trackingImpression.totalDisplayDuration += elapsedTime
             trackingImpression.tracking = false
-            currentlyTrackingImpressions[impressionId] = trackingImpression
-        }
-    }
-
-    public func resumeImpression(impressionId: String) {
-        if var trackingImpression = currentlyTrackingImpressions[impressionId], !trackingImpression.tracking {
-            trackingImpression.startTime = Date()
-            trackingImpression.tracking = true
-            currentlyTrackingImpressions[impressionId] = trackingImpression
+            currentlyTrackingImpressions[messageId] = trackingImpression
         }
     }
 
-    public func createOrUpdateImpression(impressionId: String) {
-        if let _ = currentlyTrackingImpressions[impressionId] {
-            if let index = session?.impressions.firstIndex(where: { $0.messageId == impressionId }) {
+    public func startImpression(messageId: String) {
+        if let _ = currentlyTrackingImpressions[messageId] {
+            if let index = session?.impressions.firstIndex(where: { $0.messageId == messageId }) {
                 session?.impressions[index].displayCount += 1
             }
         } else {
-            let newImpression = IterableEmbeddedImpression(messageId: impressionId, displayCount: 1, displayDuration: 0)
+            let newImpression = IterableEmbeddedImpression(messageId: messageId, displayCount: 1, displayDuration: 0)
             session?.impressions.append(newImpression)
-            currentlyTrackingImpressions[impressionId] = (totalDisplayDuration: 0, startTime: Date(), tracking: true)
+            currentlyTrackingImpressions[messageId] = (totalDisplayDuration: 0, startTime: Date(), tracking: true)
         }
         
-        resumeImpression(impressionId: impressionId)
+        resumeImpression(messageId: messageId)
+    }
+    
+    private func resumeImpression(messageId: String) {
+        if var trackingImpression = currentlyTrackingImpressions[messageId], !trackingImpression.tracking {
+            trackingImpression.startTime = Date()
+            trackingImpression.tracking = true
+            currentlyTrackingImpressions[messageId] = trackingImpression
+        }
     }
     
     private func updateDisplayDurations() {
```

Comment:
I'd like to keep the updateDisplayDurations() method since the loop would need to remain either way (it's private anyway, so I don't think it matters). We need the separate object `currentlyTrackingImpressions` to record the following about each of the current impressions: a) total display duration (which is the aggregate of each impression, b) the last start time, and c) whether or not the impression is currently being tracked, which we use to determine if the impression should be paused and the total display duration should be updated (e.g., pauseImpression()) - which is calculated by subtracting the last start time from current time.

Without the separate `currentlyTrackingImpressions` object, we would have to embed this tracking information directly into the `IterableEmbeddedSession` or `IterableEmbeddedImpression` objects, which would complicate their design and responsibilities. Currently, these objects serve the purpose of providing a record of the session, rather than actively managing the tracking of impressions in real-time. Mixing these concerns could potentially make the code more complex and harder to maintain.

RE: the efficiency of the loop in updateDisplayDurations(), it's true that it adds extra iterations at the end of the session, but the number of iterations is directly proportional to the number of unique message IDs, and unless we're dealing with an extremely large number of unique message IDs (all of which, the user has observed), the performance impact should be minimal.